src/Features/Core/Completion/CompletionListProvider.cs

```diff
@@ -1,28 +1,27 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
-using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Options;
 using Microsoft.CodeAnalysis.Text;
 
-namespace Microsoft.CodeAnalysis.Completion.Providers
+namespace Microsoft.CodeAnalysis.Completion
 {
     /// <summary>
     /// Extensibility interface for clients who want to participate in completion inside an editor.
     /// </summary>
-    internal interface ICompletionProvider
+    internal abstract class CompletionListProvider
     {
         /// <summary>
-        /// Returns a CompletionItemGroup for the specified position in the document.
+        /// Implement to register the items and other details for a <see cref="CompletionList"/>
         /// </summary>
-        Task<CompletionItemGroup> GetGroupAsync(Document document, int position, CompletionTriggerInfo triggerInfo, CancellationToken cancellationToken = default(CancellationToken));
+        public abstract Task RegisterCompletionListAsync(CompletionListContext context);
```

Comment:
That sounds like the right call to me.  Let's keep them super consistent to start with.  But then let's have a list of things for us to consider as a team as the best way to have this sort of pattern throughout our extensibility points.

Coming up with a good name for "Just do the work" would be on that list of me.  As would "Should we always use a struct, even if it's an area where we only need to create a single guy?"

Thanks!